### ISSUE TITLE
[BUG] Breaking Components

### ISSUE THREAD

**Issue by RukmaPratista** [stevahnes/figma-plugins#11]
Hi, we like your PageClone plugin, really helpful for archiving. Found a bug today though.

After I clone a page, I realized that some components are breaking. Not sure why. Please check the attached images.

Also, I agree with this https://github.com/stevahnes/figma-plugins/issues/8, detaching style as well would be great.

Before:
![01  Dashboard](https://user-images.githubusercontent.com/5291114/100428495-9d8f3c80-30c6-11eb-9bce-192a4c1b5d2d.jpg)

After:
![01  Dashboard-error](https://user-images.githubusercontent.com/5291114/100428481-99631f00-30c6-11eb-91e8-24a06aea68dd.jpg)

**Comment by stevahnes** [comment]
Hi Rukma! Thank you for highlighting this. Seems like deep nested components/instances that are hidden are not getting hidden. I will try to look into this ASAP :)

**Comment by RukmaPratista** [comment]
Terima kasih. 👍

**Comment by stevahnes** [comment]
Hi @RukmaPratista, I think I found the bug. I have released a new version. Please help test ya :) It should be version 2.2.2.

**Comment by stevahnes** [comment]
Sorry fixed another bug that might impact you. Please use 2.2.3 😊

**Comment by RukmaPratista** [comment]
Whoa, that's quick. Thanks. Anyway, I've tried the latest version, and I've got another problem. Check these screenshots, I've marked the differences.

Before:
![01 Source - A1 - Source-Error](https://user-images.githubusercontent.com/5291114/100446363-40ed4b00-30e1-11eb-83f0-a082686e9fa7.jpg)

After:
![01 Source - A1 - Source](https://user-images.githubusercontent.com/5291114/100446373-4480d200-30e1-11eb-8e24-3c972db327e7.jpg)

**Comment by stevahnes** [comment]
Hi @RukmaPratista, seems like the "clone" function does not clone specific node's properties. I have added them into the logic. Can you verify again? The version now is 2.2.4 :) If you notice anything wrong with texts, also please let me know :)

**Comment by stevahnes** [comment]
Hi @RukmaPratista have you tried the latest version?

**Comment by RukmaPratista** [comment]
Hi @stevahnes, sorry, I've tried it the other day but forgot to tell you the result. Unfortunately, I still get the same error from last time. But I'm not too sure, do I need to restart Figma to get the updated plugin? If I remembered it correctly, I didn't restart Figma that time to test 2.2.4. I'll let you know after I test it once more later.

**Comment by stevahnes** [comment]
Hi @RukmaPratista, I see.. Well I think it might take me a while to debug this one again. If it's okay with you, can you help provide the following information:

1. Is the button an instance or a main component?
2. Can you screenshot for me the righthand property panel of the button?

I will need to get to the bottom of why the clone function doesn't carry over the modifications made for the buttons :) Thank you!

**Comment by RukmaPratista** [comment]
Ok, let me answer that.

1. It's an instance.
2. Screenshots are as follows.

Before:
<img width="1920" alt="CleanShot 2020-11-30 at 21 05 46@2x" src="https://user-images.githubusercontent.com/5291114/100619640-0a693780-3350-11eb-8b55-5248a7c690f4.png">

After:
<img width="1920" alt="CleanShot 2020-11-30 at 21 08 23@2x" src="https://user-images.githubusercontent.com/5291114/100619789-35ec2200-3350-11eb-9c73-47507b4caf95.png">


One more thing, I tried to clone a page with a lot of frames. It seemed that the plugin had finished the job, but it didn't close by itself. But it's not the case after I tried it once more and only selected few frames.

It seems that you've made it to detach the styles as well. Nice!

**Comment by RukmaPratista** [comment]
Hi @stevahnes, is there any update? :)

**Comment by stevahnes** [comment]
Hi @RukmaPratista apologies I've been caught up with several trainings on my end. I will try to deliver a fix by end of next week, if that's okay with you? :)

**Comment by RukmaPratista** [comment]
Haha, that's okay my friend. It's not like I'm paying you to fix this bug. It's your plugin and you can do whatever you want with it, although my work kinda depends on it, hahaha. Anyway, if you accept a donation for this project, I'd be happy to chip in. Just let me know.

**Comment by stevahnes** [comment]
Thanks for understanding! No" I'll definitely try to fix it ASAP knowing that others' work depends on it. I am doing this for fun for now so I am happy you found a good use for it! :)

**Comment by alexcamlo** [comment]
Hi
Just wanted to let know that I also am having this issue.

In my case I'm losing borders and radius of instances inside of a variant-instance.
Just it helps pinpoint the issue.

**Comment by stevahnes** [comment]
Thank you @thewinger! I'll do my best to figure this tricky bug. Your information definitely helps! Will keep working on it!

**Comment by stevahnes** [comment]
Hi All, I am planning a dedicated session to look into this bug this coming Wednesday; will post my findings here :) Thanks again for your patience and for providing as much detail as possible!

**Comment by stevahnes** [comment]
Hi @RukmaPratista and @thewinger, I have just released PageClone 2.2.5. Please take a look and see if it solves your issue. I think Rukma's should be solved as I understood and managed to replicate the issue :)

**Comment by alexcamlo** [comment]
In my case now it brakes but in a different case.
This is the original screen
![image-orig](https://user-images.githubusercontent.com/121455/102379655-5d392500-3fc7-11eb-8bab-c745b57cf9a6.png)

The first time I used the plugin this was the result, borders disappeared
![image-borders](https://user-images.githubusercontent.com/121455/102379703-6cb86e00-3fc7-11eb-8ff4-37f6abedfcc3.png)

This is the new result, borders stay but resizing was messed up
![image-resize](https://user-images.githubusercontent.com/121455/102379737-76da6c80-3fc7-11eb-8a52-b8c2171d1bbf.png)

**Comment by stevahnes** [comment]
@thewinger can you outline to me what the node structure is? (E.g. instance --> frame --> Rectangle). A screenshot will be great if you're okay 🙂

Essentially the clone function doesn't clone everything so i need to know what's safe to copy manually and what isn't.

**Comment by alexcamlo** [comment]
In some cases it's modifying the padding numbers of the autolayout

![image](https://user-images.githubusercontent.com/121455/102383323-af7c4500-3fcb-11eb-8fff-635d4cd8e9cb.png)

![image](https://user-images.githubusercontent.com/121455/102383412-c753c900-3fcb-11eb-869d-6649e4d9c3bb.png)

And other is also modifying the resizing, apart from the padding

![image](https://user-images.githubusercontent.com/121455/102383646-139f0900-3fcc-11eb-86f4-0ab8b2fd5fc3.png)

![image](https://user-images.githubusercontent.com/121455/102383611-097d0a80-3fcc-11eb-85e3-50a9fad46117.png)

**Comment by stevahnes** [comment]
Ah okay. I left out some autolayout properties. Let me try adding them tomorrow and see if it fixes it 😊

**Comment by stevahnes** [comment]
@thewinger I just release 2.2.6. I tested it with my company's mockups that contains auto-layout and it works fine :) Please let me know if it doesn't!

**Comment by RukmaPratista** [comment]
Hi @stevahnes , thanks for working on the issues. I've tried your latest update. I have 2 issues:

1. The plugin took forever to complete and I had to terminate it prematurely. I don't think the frames are too complex, as you can see from my screenshots.

2. It added a frame or element that isn't part of my design (or part of my design but misplaced, I'm not sure)

Before: https://cln.sh/firzeq

After: https://cln.sh/7bHzej

**Comment by stevahnes** [comment]
Hi @RukmaPratista, for (2) when you terminate the plugin prematurely it stops the cloning. Hence the misplaced item, which was probably still being processed and about to be placed in the right frame. If you see the cloned frame, you will find that some of the instances are still linked to the main component.

For (1), the plugin takes long because when we detach the instance, we literally recreate the component, and then copy the individual attributes to the new clone. I have yet to find a way to do this faster, apologies! However, as long as no error is thrown, the plugin runs fine :) Figma will throw an error if something is wrong.

**Comment by alexcamlo** [comment]
@stevahnes I just tried it out and can confirm that the result is the one expected and the plugin was fast.
Thank you!

**Comment by stevahnes** [comment]
@thewinger great happy to hear that! :) It might get laggy when the frames has a lot of sub-frames and components/instances. But as long as there's no error thrown it should work fine :) Figma will throw error if something is wrong (it will pop-up as a banner below).

**Comment by stevahnes** [comment]
A note for both @thewinger and @RukmaPratista, if the plugin takes too long to run, I think Figma may throw a timeout. In that case, consider cloning the frames in batches. You can select to clone to an existing page instead of creating a new one, so cloning in batches is possible.

**Comment by RukmaPratista** [comment]
Hi @stevahnes, I made a video to show you. At first, I only chose 4 frames from total 6 frames which are quite similar in terms of complexity. I had to terminate this first attempt because it feels like forever compared to the second attempt. In the second attempt, I only chose 2 frames and it was really fast. I was expecting that the first attempt would only take twice as much time as my second attempt. 

The video duration for the first attempt is only 2:35 because I was reluctant to upload a big video file. But actually, it took a much longer time than that (and I had to end it prematurely as well).

First attempt: https://cln.sh/wWvXOq
Second attempt: https://cln.sh/FVIzsu

**Comment by stevahnes** [comment]
Hi @RukmaPratista thanks! The first video shows that something broke (the plugin shouldn't zoom into the page before the clone is completed). Are you okay with doing the following:

1. Open console (in Figma we can do that, it's in the menu)
2. Run the plugin for scenario 1 again
3. Screenshot the error thrown in the console?

I'll try to find a resource on mine that'll trigger the same error as well. Thanks!

**Comment by RukmaPratista** [comment]
Here you go:
<img width="1920" alt="CleanShot 2020-12-17 at 18 46 23@2x" src="https://user-images.githubusercontent.com/5291114/102484009-3565c200-4098-11eb-96d7-b5c24febff42.png">

**Comment by stevahnes** [comment]
Ahhhh! Okay I thought I checked for all of these! Sorry I did it this morning before work, might've missed something.. I'll check and release another version tonight. Apologies!

**Comment by stevahnes** [comment]
Hi @RukmaPratista, published 2.2.7... i didn't know that Stars, Polygons, etc can have mixed cornerRadius too. I've done the check in the code. The error should be fixed. Please let me know!

**Comment by RukmaPratista** [comment]
Hi @stevahnes, the problem still persists. However, it seems the error code is different.

<img width="1920" alt="CleanShot 2020-12-18 at 05 32 06@2x" src="https://user-images.githubusercontent.com/5291114/102551756-9fab5080-40f2-11eb-9259-78296f6e4277.png">

**Comment by stevahnes** [comment]
Hi @RukmaPratista looks like a small typo on my end but decided to do a strict number type check nonetheless. 2.2.8 is released! I hope it fixes it this time :)

**Comment by RukmaPratista** [comment]
Hahaha, this bug still persists… https://cln.sh/qCl3eU

**Comment by stevahnes** [comment]
@RukmaPratista no way!

I think it's something on the Annotation Frame. Do you mind showing what's inside that frame/component? I thought I'd handled anywhere cornerRadius pops up :(

![image](https://user-images.githubusercontent.com/29354437/102589051-6651ff80-4149-11eb-9751-e38f0c5088e8.png)

I can see that you have mixed radius for this particular component/frame (bottom left is 0 and the others are some value). Just need to ensure it's handled :)

I'd like to reconstruct that frame in my Figma if that's okay with you.

You can email the screenshot to me at stevanus.satria@gmail.com if you prefer.

**Comment by stevahnes** [comment]
Hi @RukmaPratista, I went ahead and tried all shapes that can have the corner radius property, and I went to ensure they are all not uniform (e.g. top left 10 and bottom right 20).

![image](https://user-images.githubusercontent.com/29354437/102676463-a3190780-41d8-11eb-8967-02be20ad6652.png)

And they clone just fine:

![image](https://user-images.githubusercontent.com/29354437/102676472-b0ce8d00-41d8-11eb-9c65-b9f30e99959d.png)

Will really need to know your file structure so I can try to reproduce the issue :) Thank you!

**Comment by RukmaPratista** [comment]
> I think it's something on the Annotation Frame. Do you mind showing what's inside that frame/component? I thought I'd handled anywhere cornerRadius pops up :(
> 
> ![image](https://user-images.githubusercontent.com/29354437/102589051-6651ff80-4149-11eb-9751-e38f0c5088e8.png)
> 
> I can see that you have mixed radius for this particular component/frame (bottom left is 0 and the others are some value). Just need to ensure it's handled :)

I don't think that was the case. As you can see from my previous 2 screen-captured videos, both first and second attempts tried to clone Annotation frames. The only differences between those frames are their contents: text, icon, and color.

However, I have good news. Your plugin worked flawlessly this morning! HOORAY! I'm not sure why, haha. I don't change my design or frames and I even selected all 6 frames from my testing page. 

It still feels slow for pages with lots of frames though. Some attempts to clone those pages need me to terminate the process prematurely as well. I hope you can improve it somehow. Also, if there's any way to clone the prototype settings, that would be awesome.

Anyway, thank you for your support @stevahnes, really really appreciate it. 👍👍👍 My offer still intact, if you're opening sponsorship to develop this plugin, please let me know. Don't expect too much, haha, but I hope I can be helpful in some ways.

**Comment by stevahnes** [comment]
@RukmaPratista you just made my day and maybe week! Glad that it works now 😄 
I will ask some of the other Plugin Creators on feedback to speed up the cloning process. Let's see!
Will also try to take a look at the prototype settings feature. 

I will close this ticket for now ya. If anything else comes up, please do not hesitate to open another ticket. @thewinger same goes to you!

Thank you for the offer as well. If I do end up setting up some preferred way to get sponsorship, I'll let you know? Haha... But I am just happy the Plugin helps the wider Figma community. Thank you again, and good day!

**Comment by RukmaPratista** [comment]
> Thank you for the offer as well. If I do end up setting up some preferred way to get sponsorship, I'll let you know? Haha... But I am just happy the Plugin helps the wider Figma community. Thank you again, and good day!

Yes, feel free to let me know, @stevahnes
